compiler/rustc_interface/src/passes.rs

```diff
@@ -321,6 +321,31 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
     )
 }
 
+fn env_var<'tcx>(tcx: TyCtxt<'tcx>, key: &'tcx OsStr) -> Option<&'tcx OsStr> {
+    let value = std::env::var_os(key);
+
+    let value_tcx = value.as_deref().map(|value| {
+        let encoded_bytes = tcx.arena.alloc_slice(value.as_encoded_bytes());
+        debug_assert_eq!(value.as_encoded_bytes(), encoded_bytes);
+        // SAFETY: The bytes came from `as_encoded_bytes`, and we assume that
+        // `alloc_slice` is implemented correctly, and passes the same bytes
+        // back (debug asserted above).
+        unsafe { OsStr::from_encoded_bytes_unchecked(encoded_bytes) }
+    });
+
+    // Also add the variable to Cargo's dependency tracking
+    //
+    // NOTE: This only works for passes run before `write_dep_info`. See that
+    // for extension points for configuring environment variables to be
+    // properly change-tracked.
+    tcx.sess.psess.env_depinfo.borrow_mut().insert((
+        Symbol::intern(&key.to_string_lossy()),
+        value.map(|value| Symbol::intern(&value.to_string_lossy())),
+    ));
```

Comment:
Maybe skip this step if the string is not UTF-8, instead of putting the incorrect values into the map?

In a similar situation `tracked_env::var` in `library\proc_macro\src\lib.rs` resets non-UTF-8 *values* into `None`, perhaps unintentionally.

Not sure what would be the best recovery strategy here.

Reply:
It would probably be better to turn `psess.env_depinfo` into `FxIndexSet<(OsString, Option<OsString>)>` and delay the recovery until `fn write_out_deps`.
Maybe use `to_string_lossy` there, just make sure the lossy values don't accidentally collide with the non-lossy ones.

Not in this PR though.

Reply:
I've changed this code to do the same as in `tracked_env::var` for now, and have opened https://github.com/rust-lang/rust/issues/133699 to track using `OsStr` in `env_depinfo`